client/services/code_evaluators/PythonPrerequisiteCheckService.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for performing pre-requisite checks on Python code snippets.
+ */
+
+tie.factory('PythonPrerequisiteCheckService', [
+  'PYTHON_STANDARD_LIBRARIES',
+  'PrerequisiteEvalResultObjectFactory',
+  function(
+      PYTHON_STANDARD_LIBRARIES,
+      PrerequisiteEvalResultObjectFactory) {
+
+    const LEFT_TRIM_PATTERN = /^\s+/;
+
+    var checkStarterCodePresent = function(starterCode, code) {
+      var starterCodeLines = starterCode.split('\n');
+      var codeLines = code.split('\n');
+      for (var i = 0; i < codeLines.length; i++) {
+        codeLines[i] = codeLines[i].trim();
+      }
+      var codeLineIndex = 0;
+      for (var i = 0; i < starterCodeLines.length; i++) {
+        var starterCodeLine = starterCodeLines[i].trim();
+        // ignore blank lines in starter code
+        if (starterCodeLine.length === 0) {
+          continue;
+        }
+        //handle "return" statements in starter code
+        if (starterCodeLine.replace(LEFT_TRIM_PATTERN,"").startsWith("return ")) {
+          while (codeLineIndex < codeLines.length) {
+            if (codeLines[codeLineIndex].replace(
+              /^\s+/,"").startsWith("return ")) {
+                break;
+            }
+            codeLineIndex++;
+          }
+        }
+        else {
+          while (codeLineIndex < codeLines.length) {
+            if (codeLines[codeLineIndex] === starterCodeLine) {
+              break;
+            }
+            codeLineIndex++;
+          }
+        }
+        if (codeLineIndex >= codeLines.length && i < starterCodeLines.length) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    var getCodeLibs = function(code) {
+      var codeLines = code.split('\n');
+      var coddLibs = [];
+      var pattern = new RegExp('^import\\ \\w+$');
+      for(var i=0; i<codeLines.length; i++) {
+        if(pattern.test(codeLines[i])) {
+          var words = codeLines[i].split(' ');
+          coddLibs.push(words[1]);
```

Comment:
You should be able to get the pattern directly from the regexp check by using capture groups. I recommend doing that instead, it's much cleaner.